ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/committer/transaction/TransactionContractKeysValidation.scala

```diff
@@ -0,0 +1,295 @@
+// Copyright (c) 2021 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package com.daml.ledger.participant.state.kvutils.committer.transaction
+
+import com.daml.ledger.participant.state.kvutils.Conversions
+import com.daml.ledger.participant.state.kvutils.DamlKvutils.{
+  DamlContractKey,
+  DamlStateKey,
+  DamlStateValue,
+}
+import com.daml.ledger.participant.state.kvutils.committer.Committer.Step
+import com.daml.ledger.participant.state.kvutils.committer.transaction.TransactionCommitter.{
+  DamlTransactionEntrySummary,
+  damlContractKey,
+  globalKey,
+}
+import com.daml.ledger.participant.state.kvutils.committer.{StepContinue, StepResult}
+import com.daml.ledger.participant.state.v1.RejectionReason
+import com.daml.lf.data.Ref.TypeConName
+import com.daml.lf.data.Time.Timestamp
+import com.daml.lf.transaction.{Node, NodeId}
+import com.daml.lf.value.Value
+import com.daml.lf.value.Value.ContractId
+
+object TransactionContractKeysValidation {
+
+  private type RawContractId = String
+
+  private[transaction] def validate(
+      transactionCommitter: TransactionCommitter
+  ): Step[DamlTransactionEntrySummary] =
+    (commitContext, transactionEntry) => {
+      val damlState = commitContext
+        .collectInputs[(DamlStateKey, DamlStateValue), Map[DamlStateKey, DamlStateValue]] {
+          case (key, Some(value)) if key.hasContractKey => key -> value
+        }
+      val contractKeyDamlStateKeysToContractIds: Map[DamlStateKey, RawContractId] =
+        damlState.collect {
+          case (k, v) if k.hasContractKey && v.getContractKeyState.getContractId.nonEmpty =>
+            k -> v.getContractKeyState.getContractId
+        }
+      val contractKeyDamlStateKeys: Set[DamlStateKey] =
+        contractKeyDamlStateKeysToContractIds.keySet
+      val contractKeysToContractIds: Map[DamlContractKey, RawContractId] =
+        contractKeyDamlStateKeysToContractIds.map(m => m._1.getContractKey -> m._2)
+
+      validateTransactionContractKeysUniquenessAndConsistency(
+        transactionCommitter,
+        commitContext.recordTime,
+        transactionEntry,
+        contractKeyDamlStateKeys,
+        contractKeysToContractIds,
+      ) match {
```

Comment:
I'd find this much easier to read if you implemented `map` and `flatMap` on `StepResult`. I think it would also allow you to flatten out the logic, so that we could more easily follow through the steps.

Reply:
Done.